slash_commands/adora-sac.js

```diff
@@ -0,0 +1,168 @@
+    try {
+        let excludedTowers = parseExcludedTowers(interaction.options.getString('excluded_towers'));
+
+        let xp = interaction.options.getInteger('xp');
+
+        // optimization: use money divided by 5 to shrink DP space as all tower/upgrade costs are divisble by 5
+        let moneyToSpendDiv5 = Math.ceil(xp / (4 * 5));
+
+        const costsDiv5ToUpgrade = new Map();
+
+        let towers = allTowers()
+        // sort towers in descending priority order, high priority should be looked at first
+        towers.sort((a, b) => (TOWER_PRIORITIES.get(b) || 0) - (TOWER_PRIORITIES.get(a) || 0));
+
+        const allUpgradeSets = allUpgradeCrosspathSets();
+
+        for (let tower of towers) {
+            let baseCost = costOfTowerUpgrade(tower, '000', 'hard');
+            let topPathCosts = cumulativeTowerUpgradePathCosts(tower, 'top_path', 'hard');
+            let midPathCosts = cumulativeTowerUpgradePathCosts(tower, 'middle_path', 'hard');
+            let bottomPathCosts = cumulativeTowerUpgradePathCosts(tower, 'bottom_path', 'hard');
+            for (let xpathSet of allUpgradeSets) {
+                const upgradeSetStr = `${tower}#${xpathSet}`;
+
+                // exclude t5 beast handlers due to merge requirements
+                // exclude monkeyopolis
+                if (
+                    !excludedTowers.has(upgradeSetStr)
+                    && (tower !== 'beast_handler' || !xpathSet.includes('5'))
+                    && (tower !== 'monkey_village' || xpathSet[2] !== '5')
+                ) {
+                    let rawCost = baseCost + topPathCosts[parseInt(xpathSet[0])] + midPathCosts[parseInt(xpathSet[1])] + bottomPathCosts[parseInt(xpathSet[2])];
+                    let processedCost = Math.ceil(rawCost / 5);
+                    // don't overwrite if higher priority present (deprioritize lower priority towers in case of cost tie)
+                    if (!costsDiv5ToUpgrade.has(processedCost)) {
+                        costsDiv5ToUpgrade.set(processedCost, upgradeSetStr);
+                    }
+                }
+            }
+        }
+
+        let upgradeCostKeys = [...costsDiv5ToUpgrade.keys()];
+        upgradeCostKeys.sort((a, b) => a - b);
+        
+        // cost of best sacrifice set is at most the cost of the smallest single upgrade set larger than target
+        let dpSize = Math.min(upgradeCostKeys.find((val) => val >= moneyToSpendDiv5) || XP_CAP_ADJUSTED, XP_CAP_ADJUSTED);
+
+        // set up DP arrays
+        let bagSize = new Array(dpSize+1);
+        bagSize.fill(Infinity);
+        bagSize[0] = 0
+        let lastItem = new Array(dpSize+1);
+        lastItem.fill(0);
+
+        for (let i = 1; i <= dpSize; ++i) {
+            for (let cost of upgradeCostKeys) {
+                if (i - cost < 0) break;
+                if (bagSize[i] > bagSize[i - cost] + 1) {
+                    bagSize[i] = bagSize[i - cost] + 1;
+                    lastItem[i] = cost;
+                }
```

Comment:
Maybe, as an optimization/follow-up (definitely not needed to merge this), this algorithm could store_all_ of the options that result in the smallest bagsize for a given `i` rather than just take the new smallest and ignore ties (not sure if the memory space becomes too big at this point though); the algorithm then in the below loop could compare all of the lastItem chains and pick the one with the best/highest priority..

Definitely more complex for just a small gain, but potentially worth it..

Reply:
No need to store all options (see https://github.com/hemisemidemipresent/cyberquincy/pull/276/commits/a8631c98ec5af48a40fa820033ba33fe924189c7)